### DISCUSSION TITLE
Merge several paginated JSON using "tags" and access all as Collection?

### DISCUSSION THREAD

**Discussion by devallibus** [11ty/eleventy#1799] [Q&A] [answered]
Hey,

I've been using 11ty for a while with tremendous success but have found a bit of a snag recently trying to do something more ambitious (for me).

I am making a price comparison website. I scrape each website I monitor and have an API available with the results, per website (using simplescraper.io) Inside the "_data" folder, I have a .js file for each website's API.

Example:

```
const Cache = require("@11ty/eleventy-cache-assets");

module.exports = async function() {
	let url = "XXXXXX";

	/* This returns a promise */
	return await Cache(url, {
		duration: "4h", // save for 4 hours
		type: "json" // we’ll parse JSON for you
	});
};
```

Now, I make a folder "website1", where I paginate over website1's JSON from the API call.

Example: 

```
---
tags: testtag
pagination:
    data: website1
    size: 1
    addAllPagesToCollections: true
---

{{ website1.title }}
```

11ty is nicely paginating the JSON, but I need to put **everything** inside a global collection. Reason is, I have to scrape by website, but I'm hoping to access all products with Nunjucks and sort by their price in a "lowest to highest" list for the consumer to choose the cheapest option.

Because my JS skills are close to none, I have tried to add the 

`tags: testtag`

To access it on index.njk like this, for example:

`{{ collections.all }}`

However, this results in the following output: 

`[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]`

Could you please help me see what I'm doing wrong here? I realize there might be a more elegant solution but like I said, I'm self taught and just barely scraping by. Thank you for your help.

**Comment by pdehaan** [comment, marked as answer]
When using pagination, I like to use the `alias` key as well to make it easier for me to remember the current pagination item. https://www.11ty.dev/docs/pagination/#aliasing-to-a-different-variable

Also, doesn't look like you're setting explicit permalinks, so I'm not sure how many files are being written out and what the structure of that is. But from your output, it looks like you have 18 items in the `all` collection. You could probably also output the `collections.testtag` variable and see if that has 18 items, or if you have multiple different websites and each one has a few different pages/collections.

Another trick I like to use is a special filter for dumping objects out into JSON strings for easier debugging.

```js
// .eleventy.js

eleventyConfig.addFilter("toJson", (value, pretty=2) => JSON.stringify(value, null, pretty));
```

That doesn't always play nice w/ collections and complex Eleventy data due to circular references, so another handy one is:

```js
// .eleventy.js

const inspect = require("util").inspect;
//...
eleventyConfig.addFilter("inspect", value => inspect(value));
```

Then you should be able to do something like this to debug the contents of each of the objects in the collections in question (versus the less helpful "[object Object]"):

```liquid
<pre>
{{ collections.testtag | inspect }}
</pre>
```

**Reply by devallibus** [reply]
Hi, thank you for your help.

When trying to test again to give you more info, I've come across another error.

I'll try to show more.

Right now I'm using some test data. So, I've downloaded this https://beer.mikestreety.co.uk/api/beers.json as a json, and lives on "_data/beerz.json"

Now, I've created the "src/beerz" folder, where I have the following "beerz.njk":

```
---
tags: testtag
pagination:
    data: beerz
    size: 1
    addAllPagesToCollections: true
permalink: "/beerz/{{ beer.title | slug }}/"
---

{{ beer.title }} is {{ beer.title }} years old
```

I added the permalink after your suggestion, and I get 

```
Problem writing Eleventy templates: (more in DEBUG output)
> Having trouble rendering njk template ./src/beerz/beerz.njk

`TemplateContentRenderError` was thrown
> (./src/beerz/beerz.njk)
  Error: slugify: string argument expected

`Template render error` was thrown:
    Template render error: (./src/beerz/beerz.njk)
      Error: slugify: string argument expected
```

If I don't use permalink, then it works and paginates 489 "beerz" 

```
Writing dist/index.html from ./src/index.njk.
Writing dist/beerz/index.html from ./src/beerz/beerz.njk.
Writing dist/beerz/1/index.html from ./src/beerz/beerz.njk.
Writing dist/beerz/2/index.html from ./src/beerz/beerz.njk.
```

Etcetera

Using `{{ collections.testtag }}` on "src/index.njk" gives me again this (now pasted in full, sorry for cropping earlier)

`[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]`

To be precise, in my index.njk I also have

```
 {% for beer in beerz %}
	{{ beer.title }}<br>
{% endfor %}
```

...which correctly outputs the name of each beer.

I'm a bit lost, I'm afraid.

Thanks a lot for your help @pdehaan !

**Reply by pdehaan** [reply]
```njk
---
tags: testtag
pagination:
    data: beerz
    size: 1
    addAllPagesToCollections: true
permalink: "/beerz/{{ beer.title | slug }}/"
---

{{ beer.title }} is {{ beer.title }} years old
```

> I added the permalink after your suggestion, and I get
> 
> Problem writing Eleventy templates: (more in DEBUG output)
> > Having trouble rendering njk template ./src/beerz/beerz.njk
>
> `TemplateContentRenderError` was thrown
> > (./src/beerz/beerz.njk)
>   Error: slugify: string argument expected

Ah yes, the error is because Eleventy doesn't know what `beer` is. You need to specify that as an `alias` in your frontmatter pagination object, and then it should work (I think). As it stands now, `beer.title` is probably evaluating to `undefined` and then the `slug` filter implodes if it doesn't get an explicit string data type. 


```diff
---
tags: testtag
pagination:
    data: beerz
    size: 1
+   alias: beer
    addAllPagesToCollections: true
permalink: "/beerz/{{ beer.title | slug }}/"
---

{{ beer.title }} is {{ beer.title }} years old
```

Then you'll probably get a bunch of pages (like /beerz/written-in-the-dust/index.html) that say:

> _Written in the Dust_ is _Written in the Dust_ years old.

**Reply by devallibus** [reply]
Hello again, thank you for the continued help.

So I've done that, and it works! Granted, if I use beer.title there are several titles matching and it implodes, but using:

```
---
tags: testtag
pagination:
    data: beerz
    alias: beer
    size: 1
    addAllPagesToCollections: true
permalink: "/beerz/{{ beer.number }}/"
---

{{ beer.title }} is {{ beer.title }} years old
```

It does work


```
Writing dist/beerz/491/index.html from ./src/beerz/beerz.njk.
Writing dist/beerz/490/index.html from ./src/beerz/beerz.njk.
Writing dist/beerz/489/index.html from ./src/beerz/beerz.njk.
```

etcetera

**It still won't play nice with the testtag, as {{ collections.testtag }} still gives me the [object Object] list.**

Now, I have added what you've suggested. My .eleventy.js is now like this:

```
module.exports = (eleventyConfig) => {
    eleventyConfig.addPassthroughCopy('./src/css/styles.css')

    const inspect = require("util").inspect;
    eleventyConfig.addFilter("inspect", value => inspect(value));

    return {
        markdownTemplateEngine: 'njk',
        dataTemplateEngine: 'njk',
        htmlTemplateEngine: 'njk',
        dir: {
            input: 'src',
            output: 'dist',
        },
    }
}
```

And added

```
 <pre>
	{{ collections.testtag | inspect }}
</pre>
```

to index.njk

Now, the output is massive so I'll crop it eventually:


```
[
  {
    template: Template {
      inputPath: './src/beerz/beerz.njk',
      inputDir: 'src',
      parsed: [Object],
      extraOutputSubdirectory: '',
      outputDir: 'dist',
      _extensionMap: [EleventyExtensionMap],
      linters: [],
      transforms: [],
      plugins: {},
      templateData: [TemplateData],
      paginationData: [Object],
      isVerbose: true,
      isDryRun: false,
      writeCount: 0,
      skippedCount: 0,
      wrapWithLayouts: true,
      fileSlug: [TemplateFileSlug],
      fileSlugStr: 'beerz',
      filePathStem: '/beerz/beerz',
      _config: [Object],
      _templateRender: [TemplateRender],
      inputContent: '---\r\n' +
        'tags: testtag\r\n' +
        'pagination:\r\n' +
        '    data: beerz\r\n' +
        '    alias: beer\r\n' +
        '    size: 1\r\n' +
        '    addAllPagesToCollections: true\r\n' +
        'permalink: "/beerz/{{ beer.number }}/"\r\n' +
        '---\r\n' +
        '\r\n' +
        '{{ beer.title }} is {{ beer.title }} years old',
      frontMatter: [Object],
      dataCache: [Object],
      computedData: [ComputedData]
    },
    inputPath: './src/beerz/beerz.njk',
    fileSlug: 'beerz',
    filePathStem: '/beerz/beerz',
    data: {
      beerz: [Array],
      data: [Array],
      site: [Object],
      all: [Object],
      beers: [Array],
      pkg: [Object],
      tags: [Array],
      pagination: [Object],
      permalink: '/beerz/{{ beer.number }}/',
      page: [Object],
      beer: [Object],
      collections: [Object]
    },
    date: 2021-05-28T20:31:50.666Z,
    pageNumber: 0,
    outputPath: 'dist/beerz/491/index.html',
    url: '/beerz/491/',
    templateContent: [Getter/Setter],
    _templateContent: '\r\nRanger is Ranger years old'
  },
  {
    template: Template {
      inputPath: './src/beerz/beerz.njk',
      inputDir: 'src',
      parsed: [Object],
      extraOutputSubdirectory: '',
      outputDir: 'dist',
      _extensionMap: [EleventyExtensionMap],
      linters: [],
      transforms: [],
      plugins: {},
      templateData: [TemplateData],
      paginationData: [Object],
      isVerbose: true,
      isDryRun: false,
      writeCount: 0,
      skippedCount: 0,
      wrapWithLayouts: true,
      fileSlug: [TemplateFileSlug],
      fileSlugStr: 'beerz',
      filePathStem: '/beerz/beerz',
      _config: [Object],
      _templateRender: [TemplateRender],
      inputContent: '---\r\n' +
        'tags: testtag\r\n' +
        'pagination:\r\n' +
        '    data: beerz\r\n' +
        '    alias: beer\r\n' +
        '    size: 1\r\n' +
        '    addAllPagesToCollections: true\r\n' +
        'permalink: "/beerz/{{ beer.number }}/"\r\n' +
        '---\r\n' +
        '\r\n' +
        '{{ beer.title }} is {{ beer.title }} years old',
      frontMatter: [Object],
      dataCache: [Object],
      computedData: [ComputedData]
    },
    inputPath: './src/beerz/beerz.njk',
    fileSlug: 'beerz',
    filePathStem: '/beerz/beerz',
    data: {
      beerz: [Array],
      data: [Array],
      site: [Object],
      all: [Object],
      beers: [Array],
      pkg: [Object],
      tags: [Array],
      pagination: [Object],
      permalink: '/beerz/{{ beer.number }}/',
      page: [Object],
      beer: [Object],
      collections: [Object]
    },
    date: 2021-05-28T20:31:50.666Z,
    pageNumber: 1,
    outputPath: 'dist/beerz/490/index.html',
    url: '/beerz/490/',
    templateContent: [Getter/Setter],
    _templateContent: '\r\nWritten in the Dust is Written in the Dust years old'
  },
  {
    template: Template {
      inputPath: './src/beerz/beerz.njk',
      inputDir: 'src',
      parsed: [Object],
      extraOutputSubdirectory: '',
      outputDir: 'dist',
      _extensionMap: [EleventyExtensionMap],
      linters: [],
      transforms: [],
      plugins: {},
      templateData: [TemplateData],
      paginationData: [Object],
      isVerbose: true,
      isDryRun: false,
      writeCount: 0,
      skippedCount: 0,
      wrapWithLayouts: true,
      fileSlug: [TemplateFileSlug],
      fileSlugStr: 'beerz',
      filePathStem: '/beerz/beerz',
      _config: [Object],
      _templateRender: [TemplateRender],
      inputContent: '---\r\n' +
        'tags: testtag\r\n' +
        'pagination:\r\n' +
        '    data: beerz\r\n' +
        '    alias: beer\r\n' +
        '    size: 1\r\n' +
        '    addAllPagesToCollections: true\r\n' +
        'permalink: "/beerz/{{ beer.number }}/"\r\n' +
        '---\r\n' +
        '\r\n' +
        '{{ beer.title }} is {{ beer.title }} years old',
      frontMatter: [Object],
      dataCache: [Object],
      computedData: [ComputedData]
    },
    inputPath: './src/beerz/beerz.njk',
    fileSlug: 'beerz',
    filePathStem: '/beerz/beerz',
    data: {
      beerz: [Array],
      data: [Array],
      site: [Object],
      all: [Object],
      beers: [Array],
      pkg: [Object],
      tags: [Array],
      pagination: [Object],
      permalink: '/beerz/{{ beer.number }}/',
      page: [Object],
      beer: [Object],
      collections: [Object]
    },
    date: 2021-05-28T20:31:50.666Z,
    pageNumber: 2,
    outputPath: 'dist/beerz/489/index.html',
    url: '/beerz/489/',
    templateContent: [Getter/Setter],
    _templateContent: '\r\nParade is Parade years old'
  },
  {
    template: Template {
      inputPath: './src/beerz/beerz.njk',
      inputDir: 'src',
      parsed: [Object],
      extraOutputSubdirectory: '',
      outputDir: 'dist',
      _extensionMap: [EleventyExtensionMap],
      linters: [],
      transforms: [],
      plugins: {},
      templateData: [TemplateData],
      paginationData: [Object],
      isVerbose: true,
      isDryRun: false,
      writeCount: 0,
      skippedCount: 0,
      wrapWithLayouts: true,
      fileSlug: [TemplateFileSlug],
      fileSlugStr: 'beerz',
      filePathStem: '/beerz/beerz',
      _config: [Object],
      _templateRender: [TemplateRender],
      inputContent: '---\r\n' +
        'tags: testtag\r\n' +
        'pagination:\r\n' +
        '    data: beerz\r\n' +
        '    alias: beer\r\n' +
        '    size: 1\r\n' +
        '    addAllPagesToCollections: true\r\n' +
        'permalink: "/beerz/{{ beer.number }}/"\r\n' +
        '---\r\n' +
        '\r\n' +
        '{{ beer.title }} is {{ beer.title }} years old',
      frontMatter: [Object],
      dataCache: [Object],
      computedData: [ComputedData]
    },
    inputPath: './src/beerz/beerz.njk',
    fileSlug: 'beerz',
    filePathStem: '/beerz/beerz',
    data: {
      beerz: [Array],
      data: [Array],
      site: [Object],
      all: [Object],
      beers: [Array],
      pkg: [Object],
      tags: [Array],
      pagination: [Object],
      permalink: '/beerz/{{ beer.number }}/',
      page: [Object],
      beer: [Object],
      collections: [Object]
    },
    date: 2021-05-28T20:31:50.666Z,
    pageNumber: 3,
    outputPath: 'dist/beerz/488/index.html',
    url: '/beerz/488/',
    templateContent: [Getter/Setter],
    _templateContent: '\r\nFor Your Eyes Only is For Your Eyes Only years old'
  },
  {
    template: Template {
      inputPath: './src/beerz/beerz.njk',
      inputDir: 'src',
      parsed: [Object],
      extraOutputSubdirectory: '',
      outputDir: 'dist',
      _extensionMap: [EleventyExtensionMap],
      linters: [],
      transforms: [],
      plugins: {},
      templateData: [TemplateData],
      paginationData: [Object],
      isVerbose: true,
      isDryRun: false,
      writeCount: 0,
      skippedCount: 0,
      wrapWithLayouts: true,
      fileSlug: [TemplateFileSlug],
      fileSlugStr: 'beerz',
      filePathStem: '/beerz/beerz',
      _config: [Object],
      _templateRender: [TemplateRender],
      inputContent: '---\r\n' +
        'tags: testtag\r\n' +
        'pagination:\r\n' +
        '    data: beerz\r\n' +
        '    alias: beer\r\n' +
        '    size: 1\r\n' +
        '    addAllPagesToCollections: true\r\n' +
        'permalink: "/beerz/{{ beer.number }}/"\r\n' +
        '---\r\n' +
        '\r\n' +
        '{{ beer.title }} is {{ beer.title }} years old',
      frontMatter: [Object],
      dataCache: [Object],
      computedData: [ComputedData]
    },
    inputPath: './src/beerz/beerz.njk',
    fileSlug: 'beerz',
    filePathStem: '/beerz/beerz',
    data: {
      beerz: [Array],
      data: [Array],
      site: [Object],
      all: [Object],
      beers: [Array],
      pkg: [Object],
      tags: [Array],
      pagination: [Object],
      permalink: '/beerz/{{ beer.number }}/',
      page: [Object],
      beer: [Object],
      collections: [Object]
    },
    date: 2021-05-28T20:31:50.666Z,
    pageNumber: 4,
    outputPath: 'dist/beerz/487/index.html',
    url: '/beerz/487/',
    templateContent: [Getter/Setter],
    _templateContent: '\r\nDiceman is Diceman years old'
  },
  {
    template: Template {
      inputPath: './src/beerz/beerz.njk',
      inputDir: 'src',
      parsed: [Object],
      extraOutputSubdirectory: '',
      outputDir: 'dist',
      _extensionMap: [EleventyExtensionMap],
      linters: [],
      transforms: [],
      plugins: {},
      templateData: [TemplateData],
      paginationData: [Object],
      isVerbose: true,
      isDryRun: false,
      writeCount: 0,
      skippedCount: 0,
      wrapWithLayouts: true,
      fileSlug: [TemplateFileSlug],
      fileSlugStr: 'beerz',
      filePathStem: '/beerz/beerz',
      _config: [Object],
      _templateRender: [TemplateRender],
      inputContent: '---\r\n' +
        'tags: testtag\r\n' +
        'pagination:\r\n' +
        '    data: beerz\r\n' +
        '    alias: beer\r\n' +
        '    size: 1\r\n' +
        '    addAllPagesToCollections: true\r\n' +
        'permalink: "/beerz/{{ beer.number }}/"\r\n' +
        '---\r\n' +
        '\r\n' +
        '{{ beer.title }} is {{ beer.title }} years old',
      frontMatter: [Object],
      dataCache: [Object],
      computedData: [ComputedData]
    },
    inputPath: './src/beerz/beerz.njk',
    fileSlug: 'beerz',
    filePathStem: '/beerz/beerz',
    data: {
      beerz: [Array],
      data: [Array],
      site: [Object],
      all: [Object],
      beers: [Array],
      pkg: [Object],
      tags: [Array],
      pagination: [Object],
      permalink: '/beerz/{{ beer.number }}/',
      page: [Object],
      beer: [Object],
      collections: [Object]
    },
    date: 2021-05-28T20:31:50.666Z,
    pageNumber: 5,
    outputPath: 'dist/beerz/486/index.html',
    url: '/beerz/486/',
    templateContent: [Getter/Setter],
    _templateContent: '\r\nDIPA is DIPA years old'
  },
```

I have the dubious privilege of telling you that I don't really understand what I'm looking at here.

Any more pointers highly appreciated, thank you!

**Reply by pdehaan** [reply]
> Granted, if I use `beer.title` there are several titles matching and it implodes

Yes, `permalink`s must be unique. Eleventy protects you from having 8 different files trying to write to the exact same output file (ie: /beers/index.html). So with our current `permalink` of "/beerz/{{ beer.number }}/", we'll see some errors of multiple beers have the same title since you'll get slug collisions.

> It still won't play nice with the testtag, as {{ collections.testtag }} still gives me the [object Object] list.

Yep. `collections.testtag` is an array of 490ish objects, so what you're seeing a toString()ed version of that. But, like you discovered, if we use a helpful `toJson` or `inspect` style filter, we can see the complex, nested objects in their complex, nested glory (or horror, depending on how much fun you're currently having).

> I have the dubious privilege of telling you that I don't really understand what I'm looking at here.

It's just a lot of plumbing and everything that goes into rendering a template, whether its the template `inputPath`, a config `inputDir`, the current template's `fileSlug`, raw template input, current template's `frontMatter` as an object, global data, package.json contents, various collection data, the current pagination page number, etc. All out there in it's glory. Then repeated 489 more times. I'm sure there's a Nunjucks filter we could use to only dump a single object, but sometimes 15MB of debugging info is fun to browse. `{{ collections.testtag | first | inspect }}` might work. Maybe not. :shrug:

Not sure what the next steps are. We can paginate over a sample data file of beers, so it looks like pagination syntax is correct and we can generate unique permalinks.
I'm not sure what `website1` looks like (or if its a data file or collection name). This might work. 

```njk
---
tags: testtag
pagination:
    data: website1
    alias: site
    size: 1
    addAllPagesToCollections: true
# But this won't work if the loop iteration's `title` isn't unique.
permalink: "/{{ site.title | slug }}/"
---

<pre>
{{ site | inspect }}
</pre>
```

Do you have a public repo somewhere I can look at? Might make it easier to group debug.

**Reply by devallibus** [reply]
Thanks again!

Yes, I have published it now. Bear in mind it's just a random repo I picked up for testing.

I've removed for speed my .js calling the scraper API as I'd have to secure the key just in case some random bot decides to nuke my allowance, haha. However, the beerz.json is virtually identical to what I get from said API, and beers.js is identical save for the url ofc.

Here is the link:
https://github.com/debatz/datatest

Using your suggestion

```
<pre>
{{ site | inspect }}
</pre>
```

It does output perfectly 

```
    {
  id: '1395886658666138062',
  date: '2016-12-01',
  code: 'BNfL4U0gInO',
  has_many: 0,
  is_private: 0,
  created_at: '2020-02-24 14:43:29',
  updated_at: '2021-05-19 02:02:51',
  title: 'Mad Goose',
  brewery: 'Purity Brewing Company',
  description: "A light, hoppy beer. Could easily sit in a pub and drink several pints - not really the kind of beer you'd sip while watching tv at home though.",
  hashtags: '#bitter #ale #pint #drinking #thirstythirsday #christmasale #adventcalendar',
  rating: '7',
  number: 1,
  like_count: 11,
  comment_count: 0
}
```

in each pagination link.

Seems to work perfectly when paginating but maybe parses things into the tag collection in an unexpected manner?

I'm CET (UTC+2) so I have to stop for a couple of hours, but thank you again for you fast help, it's greatly appreciated!

**Reply by pdehaan** [reply]
Possibly something like this will work:

```njk
<pre>
{{ collections.testtag | first | inspect | safe }}
</pre>

<h1>Posts</h1>
<ol>
{% for item in collections.testtag %}
  <li>{{ item.data.beer.title | safe }}</li>
	<!-- item.data: {{ item.data | keys | safe }} -->
	<!-- item.data.beer: {{ item.data.beer | keys | safe }} -->
{% endfor %}
</ol>
```

So we have
- `item` &mdash; Our collection item for the current loop iteration.
  - `data` &mdash; The frontmatter data object.
    - `beer` &mdash; The pagination alias.
      - `title` &mdash; The `title` property from the current pagination item.

It feels kind of weird since it's a collection built using a data file and pagination and an alias that we're trying to loop over, so it had an unexpected `.beer` property.


Where my custom `keys` filter looks like:

```js
const inspect = require("util").inspect;
eleventyConfig.addFilter("inspect", value => inspect(value));
eleventyConfig.addFilter("keys", obj => JSON.stringify(Object.keys(obj).sort()));
```

### Output

```html
<li>Ranger</li>
<!-- item.data: ["beer","beers","beerz","collections","page","pagination","permalink","pkg","site","tags"] -->
<!-- item.data.beer: ["brewery","code","comment_count","created_at","date","description","has_many","hashtags","id","is_private","like_count","number","rating","title","updated_at"] -->

<li>Written in the Dust</li>
<!-- item.data: ["beer","beers","beerz","collections","page","pagination","permalink","pkg","site","tags"] -->
<!-- item.data.beer: ["brewery","code","comment_count","created_at","date","description","has_many","hashtags","id","is_private","like_count","number","rating","title","updated_at"] -->

...
```

**Reply by devallibus** [reply]
It works!

I had an inkling it would be some kind of unexpected property but didn't know what to do with it. Tried several times but there is no way in hell I'd make those filters.

To be completely honest with you, your custom filters go a bit over my head.

Thank you very, very much for your help.

I'm going to keep working on this and will update this thread. Possibly should make a write-up somewhere as I haven't seen this particular need of tagging a paginated json anywhere quite in this way?

I'll come back and give a full post-mortem in case someone else finds it interesting in the future.

Thanks a lot @pdehaan I'll update shortly.

**Reply by pdehaan** [reply]
I was thinking of bundling those debugging filters into a custom 11ty plugin last night since I tend to recreate them often for debugging... Here's the rough README.md snippet I have for documentation so far (mostly just links to the relevant documentation):

1. `inspect` &mdash; Wrapper for Node's native [`util.inspect()` method](https://nodejs.org/api/util.html#util_util_inspect_object_options). Useful if the object you're debugging might have circular references, or non-JSON-stringify-able data.
1. `json` &mdash; Wrapper for JavaScript's [`JSON.stringify()` method](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify). This filter takes one optional argument which is a string or number value to use for indentation, if you want pretty printed JSON objects.
1. `keys` &mdash; Wrapper for JavaScript's [`Object.keys()` method](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object/keys). This filter will also sort the returned array of key names for the specified object.

**Reply by devallibus** [reply]
That'd be nice!

11ty's data manipulation is one of its greatests advantages, and it is likely that after this project I'll be converting a 8K+ SKUs Woocommerce ecommerce into a Prismic + 11ty + Snipcart/Ecwid/CommerceJS (under evaluation) stack.

This is to say that I'm going to try to ELI5 11ty as much for me as for others and make my own "public documentation for dummies" in spanish (seems like I'm the only one, or near the only one, interested in 11ty for the moment here) so I'd be very interested in seeing that plugin.

Thanks a lot and I'll come back to update this when I get it working.

**Reply by pdehaan** [reply]
Just pushed to GitHub and published to npm: https://www.npmjs.com/package/eleventy-plugin-debug
